### PR TITLE
improvement: workflow engine data source extension

### PR DESCRIPTION
**Pull request by vrancurel**
Monitors a ZK specific path for new ephermeral nodes.
Those nodes contain a JSON object defining a filter descriptor.
Filter descriptors are added/removed when ephemeral nodes are created/deleted.
Unit tests.

### CODE REVIEW

**Comment by vrancurel** [comment]
remove it.only() in extension tests

File: extensions/workflowEngineDataSource/WorkflowEngineDataSourceQueuePopulator.js [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+    _getFilterDescriptorsLength() {
+        return Object.keys(this.filterDescriptors).length;
+    }
+
+    // for tests
+    _deleteAllFilterDescriptors() {
+        this.filterDescriptors = [];
+    }
+
+    // check if zkClient exists and is connected before
```
Comment (jonathan-gramain) [inline extensions/workflowEngineDataSource/WorkflowEngineDataSourceQueuePopulator.js]: You can use jsdoc style like (valid for all functions):
```
/**
 * check if ...
 *
 * @return {undefined}
 */
```

File: extensions/workflowEngineDataSource/WorkflowEngineDataSourceQueuePopulator.js [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+    // (for timer)
+    __updateFilterDescriptors() {
```
Comment (jonathan-gramain) [inline extensions/workflowEngineDataSource/WorkflowEngineDataSourceQueuePopulator.js]: I would opt for a more meaningful name than having two leading underscores to distinguish it with the following function.

What about `_updateFilterDescriptorsTimer()`?

File: extensions/workflowEngineDataSource/WorkflowEngineDataSourceQueuePopulator.js [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+        }
+        if (this.zkClient.getState !== NodeZookeeperClient.State.CONNECTED) {
+            this.log.debug('WEDSQP: zk not connected');
+            return;
+        }
+        this._updateFilterDescriptors();
+    }
+
+    // update the filter descriptor table acc/to new information
+    // cb is optional
```
Comment (jonathan-gramain) [inline extensions/workflowEngineDataSource/WorkflowEngineDataSourceQueuePopulator.js]: in jsdoc that could be: `@param {function} [cb] - callback`

File: extensions/workflowEngineDataSource/WorkflowEngineDataSourceQueuePopulator.js [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+                            error: err,
+                        });
+                    if (cb) {
+                        return cb(err);
+                    }
+                    return undefined;
+                }
+                this.log.debug(
+                    `WEDSQP: reading children data ${children.length}`);
+                async.times(children.length, (i, next) => {
```
Comment (jonathan-gramain) [inline extensions/workflowEngineDataSource/WorkflowEngineDataSourceQueuePopulator.js]: a more idiosynchratic function to do the same: `async.map(children, (child, next) => ..., (err, results) => {...})`

You might want to use `mapSeries` or `mapLimit` to control parallelism better.

File: extensions/workflowEngineDataSource/WorkflowEngineDataSourceQueuePopulator.js [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+                    }
+                    this.log.debug('WEDSQP: processing children data');
+                    const _children = {};
+                    // check for new children
+                    Object.values(results).forEach(value => {
+                        const fd = JSON.parse(value);
+                        const fdName = this._getHashString(
+                            fd.workflowId, fd.workflowVersion);
+                        _children[fdName] = true;
+                        if (!Object.prototype.hasOwnProperty.call(
```
Comment (jonathan-gramain) [inline extensions/workflowEngineDataSource/WorkflowEngineDataSourceQueuePopulator.js]: What about `if (!this.filterDescriptors[fdName]) ...` ?

File: extensions/workflowEngineDataSource/WorkflowEngineDataSourceQueuePopulator.js [hunk 6]
```diff
@@ -0,0 +1,376 @@
...
+                                        method:
+                                        'WEDSQP._updateFilterDescriptors',
+                                        error: errors.InvalidArgument,
+                                    });
+                            }
+                        }
+                    });
+                    // check for missing children
+                    const keys = Object.keys(this.filterDescriptors);
+                    for (let i = 0; i < keys.length; i++) {
```
Comment (jonathan-gramain) [inline extensions/workflowEngineDataSource/WorkflowEngineDataSourceQueuePopulator.js]: Can be simpler to use `keys.forEach(key => ...)`

File: extensions/workflowEngineDataSource/WorkflowEngineDataSourceQueuePopulator.js [hunk 7]
```diff
@@ -0,0 +1,376 @@
...
+            if (doNotSetWatcher) {
+                return cb();
+            } else {
+                return this._setZkWatcher(cb);
+            }
+        });
+    }
+
+    // get log helper
+    _gLH(fd) {
```
Comment (jonathan-gramain) [inline extensions/workflowEngineDataSource/WorkflowEngineDataSourceQueuePopulator.js]: You can call this function `_getLogHelper()` :wink:

File: extensions/workflowEngineDataSource/WorkflowEngineDataSourceQueuePopulator.js [hunk 8]
```diff
@@ -0,0 +1,376 @@
...
+            return;
+        }
+
+        // iterate over current filter descriptors
+        const keys = Object.keys(this.filterDescriptors);
+        for (let i = 0; i < keys.length; i++) {
+            const fd = this.filterDescriptors[keys[i]];
+            const _gLH = this._gLH(fd);
+
+            const bucket = queueEntry.getBucket();
```
Comment (jonathan-gramain) [inline extensions/workflowEngineDataSource/WorkflowEngineDataSourceQueuePopulator.js]: You could pre-assign those variables outside the `for` loop.

File: tests/unit/workflowEngineDataSource/WorkflowEngineDataSourceQueuePopulator.spec.js [hunk 9]
```diff
@@ -0,0 +1,364 @@
...
+        wedsqp = new WorkflowEngineDataSourceQueuePopulatorMock(params);
+    });
+
+    afterEach(() => {
+        wedsqp.resetState();
+        wedsqp._deleteAllFilterDescriptors();
+        wedsqp.zkClient._resetState();
+    });
+
+    it.only('should validate filter descriptor', () => {
```
Comment (jonathan-gramain) [inline tests/unit/workflowEngineDataSource/WorkflowEngineDataSourceQueuePopulator.spec.js]: you should remove all the `.only`

**Review by jonathan-gramain — APPROVED** [review]

**Review by bennettbuchanan — APPROVED** [review]

**Comment by vrancurel** [comment]
@bert-e approve

**Pull request merged**